numpy/core/src/common/npy_argparse.c

```diff
@@ -0,0 +1,421 @@
+    cache->nrequired = nrequired;
+    cache->nkwarg = nkwargs;
+
+    /* NULL kw_strings for easier cleanup (and NULL termination) */
+    memset(cache->kw_strings, 0, sizeof(PyObject *) * (nkwargs + 1));
+
+    va_copy(va, va_orig);
+    for (int i = 0; i < nargs; i++) {
+        /* Advance through non-kwargs, which do not require setup. */
+        char *name = va_arg(va, char *);
+        va_arg(va, convert *);
+        va_arg(va, void *);
+
+        if (i >= npositional_only) {
+            int i_kwarg = i - npositional_only;
+            cache->kw_strings[i_kwarg] = PyUString_InternFromString(name);
+            if (cache->kw_strings[i_kwarg] == NULL) {
+                va_end(va);
+                goto error;
+            }
+        }
+    }
+    va_end(va);
+    return 0;
+
+error:
+    for (int i = 0; i < nkwargs; i++) {
+        Py_XDECREF(cache->kw_strings[i]);
+    }
+    return -1;
+}
+
+static int
+raise_incorrect_number_of_positional_args(const char *funcname,
+        const _NpyArgParserCache *cache, Py_ssize_t len_args)
+{
+    if (cache->npositional == cache->nrequired) {
+        PyErr_Format(PyExc_TypeError,
+                "%s() takes %d positional arguments but %zd were given",
+                funcname, cache->npositional, len_args);
+    }
+    else {
+        PyErr_Format(PyExc_TypeError,
+                "%s() takes from %d to %d positional arguments but "
+                "%zd were given",
+                funcname, cache->nrequired, cache->npositional, len_args);
+    }
+    return 0;
+}
+
+static void
+raise_missing_argument(const char *funcname,
+        const _NpyArgParserCache *cache, int i)
+{
+    /* These errors are as precise as PyArg_*, but less than python */
+    if (i < cache->npositional_only) {
+        PyErr_Format(PyExc_TypeError,
+                "%s() missing required positional argument %d",
+                funcname, i);
+    }
+    else {
+        PyObject *kw = cache->kw_strings[i - cache->npositional_only];
+        PyErr_Format(PyExc_TypeError,
+                "%s() missing required argument '%S' (pos %d)",
+                funcname, kw, i);
+    }
+}
+
+
+/**
+ * Generic helper for argument parsing
+ *
+ * See macro version for pattern of how to use this function.
+ *
+ * @param funcname
+ * @param nrequired
+ * @param npositional
+ * @param cache
+ * @param NPY_PARAMS_PASS Macro expanding to keywords+varargs python calling
+ *                        convention, using ``METH_FASTCALL`` when available.
+ * @param ... List of arguments must be param1_name, param1_converter,
+ *            *param1_outvalue, param2_name, ..., NULL, NULL, NULL.
+ *            Where name is ``char *``, ``converter`` a python converter
+ *            function or NULL and ``outvalue`` is the ``void *`` passed to
+ *            the converter (holding the converted data or a borrowed
+ *            reference if converter is NULL).
+ *
+ * @return Returns 1 on success and 0 on failure.
```

Comment:
I'd recommend
```suggestion
 * @return Returns 0 on success and -1 on failure.
```
which is more typical for error-reporting in python - converter functions seem to be the exception not the rule.

Reply:
I generally prefer -1, I had it like this only, because it is closer to the code using `PyArg_Parse*` code, but am happy change.